codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/JsonRpcProtocolGenerator.java

```diff
@@ -0,0 +1,175 @@
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.StructureShape;
+
+/**
+ * Handles generating the aws.rest-json protocol for services.
+ *
+ * @inheritDoc
+ *
+ * @see HttpRpcProtocolGenerator
+ */
+abstract class JsonRpcProtocolGenerator extends HttpRpcProtocolGenerator {
+
+    /**
+     * Creates an AWS JSON RPC protocol generator
+     */
+    public JsonRpcProtocolGenerator() {
+        super(true);
+    }
+
+    @Override
+    protected String getOperationPath(GenerationContext context, OperationShape operation) {
+        return "/";
+    }
+
+    @Override
+    protected void writeDefaultHeaders(GenerationContext context, OperationShape operation, GoWriter writer) {
+        super.writeDefaultHeaders(context, operation, writer);
+        String target = context.getService().getId().getName() + "." + operation.getId().getName();
+        writer.write("httpBindingEncoder.SetHeader(\"X-Amz-Target\").String($S)", target);
+    }
+
+    @Override
+    protected void serializeInputDocument(GenerationContext context, OperationShape operation) {
+        GoWriter writer = context.getWriter();
+        StructureShape input = ProtocolUtils.expectInput(context.getModel(), operation);
+        String functionName = ProtocolGenerator.getDocumentSerializerFunctionName(input, getProtocolName());
+
+        writer.addUseImports(SmithyGoDependency.SMITHY_JSON);
+        writer.write("jsonEncoder := smithyjson.NewEncoder()");
+        writer.openBlock("if err := $L(input, jsonEncoder.Value); err != nil {", "}", functionName, () -> {
+            writer.write("return out, metadata, &smithy.SerializationError{Err: err}");
+        }).write("");
+
+        writer.addUseImports(SmithyGoDependency.BYTES);
+        writer.openBlock("if request, err = request.SetStream(bytes.NewReader(jsonEncoder.Bytes())); err != nil {",
+                "}", () -> {
+                    writer.write("return out, metadata, &smithy.SerializationError{Err: err}");
+                });
+    }
+
+    @Override
+    protected void generateDocumentBodyShapeSerializers(GenerationContext context, Set<Shape> shapes) {
+        JsonShapeSerVisitor visitor = new JsonShapeSerVisitor(context);
+        shapes.forEach(shape -> shape.accept(visitor));
+    }
+
+    @Override
+    protected void deserializeOutputDocument(GenerationContext context, OperationShape operation) {
+        GoWriter writer = context.getWriter();
+        StructureShape output = ProtocolUtils.expectOutput(context.getModel(), operation);
+        String functionName = ProtocolGenerator.getDocumentDeserializerFunctionName(output, getProtocolName());
+        initializeJsonDecoder(writer, "response.Body");
+        writer.write("err = $L(&output, decoder)", functionName);
+        handleDecodeError(writer, "out, metadata, ");
+    }
+
+    // TODO: this could probably be a generic utility
```

Comment:
I've seen some methods that sound like they do the same thing elsewhere, I'd +1 this in your refactoring.